go/test/endtoend/recovery/unshardedrecovery/recovery.go

```diff
@@ -185,100 +185,95 @@ func TestMainImpl(m *testing.M) {
 
+	// take first backup of value = test1
 	err := localCluster.VtctlclientProcess.ExecuteCommand("Backup", replica1.Alias)
 	assert.NoError(t, err)
 
 	backups := listBackups(t)
 	require.Equal(t, len(backups), 1)
 	assert.Contains(t, backups[0], replica1.Alias)
 
-	_, err = primary.VttabletProcess.QueryTablet("insert into vt_insert_test (msg) values ('test2')", keyspaceName, true)
-	assert.NoError(t, err)
-	cluster.VerifyRowsInTablet(t, replica1, keyspaceName, 2)
-
 	err = localCluster.VtctlclientProcess.ApplyVSchema(keyspaceName, vSchema)
 	assert.NoError(t, err)
 
 	output, err := localCluster.VtctlclientProcess.ExecuteCommandWithOutput("GetVSchema", keyspaceName)
 	assert.NoError(t, err)
 	assert.Contains(t, output, "vt_insert_test")
 
-	recovery.RestoreTablet(t, localCluster, replica2, recoveryKS1, "0", keyspaceName, commonTabletArg)
+	// restore with latest backup
+	restoreTime := time.Now().UTC()
+	recovery.RestoreTablet(t, localCluster, replica2, recoveryKS1, "0", keyspaceName, commonTabletArg, restoreTime)
 
 	output, err = localCluster.VtctlclientProcess.ExecuteCommandWithOutput("GetSrvVSchema", cell)
 	assert.NoError(t, err)
 	assert.Contains(t, output, keyspaceName)
 	assert.Contains(t, output, recoveryKS1)
 
-	err = localCluster.VtctlclientProcess.ExecuteCommand("GetSrvKeyspace", cell, keyspaceName)
-	assert.NoError(t, err)
-
 	output, err = localCluster.VtctlclientProcess.ExecuteCommandWithOutput("GetVSchema", recoveryKS1)
 	assert.NoError(t, err)
 	assert.Contains(t, output, "vt_insert_test")
 
 	cluster.VerifyRowsInTablet(t, replica2, keyspaceName, 1)
 
+	// verify that restored replica has value = test1
+	qr, err := replica2.VttabletProcess.QueryTablet("select msg from vt_insert_test where id = 1", keyspaceName, true)
+	assert.NoError(t, err)
+	assert.Equal(t, "test1", qr.Rows[0][0].ToString())
+
+	// insert new row on primary
+	_, err = primary.VttabletProcess.QueryTablet("insert into vt_insert_test (msg) values ('test2')", keyspaceName, true)
+	assert.NoError(t, err)
+	cluster.VerifyRowsInTablet(t, replica1, keyspaceName, 2)
+
 	// update the original row in primary
 	_, err = primary.VttabletProcess.QueryTablet("update vt_insert_test set msg = 'msgx1' where id = 1", keyspaceName, true)
 	assert.NoError(t, err)
 
 	// verify that primary has new value
-	qr, err := primary.VttabletProcess.QueryTablet("select msg from vt_insert_test where id = 1", keyspaceName, true)
+	qr, err = primary.VttabletProcess.QueryTablet("select msg from vt_insert_test where id = 1", keyspaceName, true)
 	assert.NoError(t, err)
 	assert.Equal(t, "msgx1", qr.Rows[0][0].ToString())
 
-	// verify that restored replica has old value
-	qr, err = replica2.VttabletProcess.QueryTablet("select msg from vt_insert_test where id = 1", keyspaceName, true)
-	assert.NoError(t, err)
-	assert.Equal(t, "test1", qr.Rows[0][0].ToString())
-
+	// take second backup of value = msgx1
 	err = localCluster.VtctlclientProcess.ExecuteCommand("Backup", replica1.Alias)
```

Comment:
We're missing validation that `replica1` has the value `"msgx1"`. The backup of `replica1` might only contains `"test1"`, which means the fact we restore and get `"test1"` does not indicate which backup we've restored.

Reply:
I should add that the way to resolve this is to run a `context.WithTimeout()` + `time.Ticker` `select` loop waiting for the data to appear in `replica1`. Our experience shows that in GitHub CI things can be unbelievably slow. I'd say give this a timeout of `10s`.

Reply:
updated!